Introduce an adapter option for bugs bunny

gausby opened the pull request:
During the initialization of bugs bunny I split out the AMQP related configs from ours; I pull out the adapter (previously called "client", which is a loaded term) and store it in the state. There is no need to pull this from the environment and it is a lot cleaner as pulling it out of the state is a simple pattern match.

I had to fiddle with the tests and the test setup to get them to pass on my system. We might have some timing issues.

Commit:
Attempt to make the unit tests pass on my machine

I had to upgrade meck to get rid of some meck compile errors, and I
needed to introduce some timeouts to beat some timing related
issues. I cannot seem to run the tests consistently still, sometimes
a handful of them fail for odd reasons.

Commit:
Make sure we do not fail silently when we cannot create & bind queue

Commit:
Keep the RabbitMQ module in the state instead of fetching from env

During the initialization of bugs bunny I split out the amqp related
configs from ours; I pull out the adapter (previously called "client",
which is a loaded term) and store it in the state. There is no need to
pull this from the environment and it is a lot cleaner as pulling it
out of the state is a simple pattern match.

sescobb27 commented on apps/repo_poller/test/poller_test.exs:
```diff
@@ -226,6 +226,7 @@ defmodule RepoPoller.PollerTest do
              assert_receive {:error, :rate_limit, retry}
              assert_receive {:error, :rate_limit, ^retry}
            end) =~ "rate limit reached for repo: rate-limit/fake retrying in 50 ms"
+    :timer.sleep 200
```
why is there a timer sleep at the end?

sescobb27 commented on apps/repo_poller/test/setup_worker_test.exs:
```diff
@@ -74,6 +74,7 @@ defmodule RepoPoller.SetupWorkerTest do
              end)
 
     worker_pid = Process.whereis(:f@k3)
+    :timer.sleep 200
```
hmmm, we need to investigate why we need a timer here if we are already "active waiting" above for workers to be created

sescobb27 commented on apps/repo_poller/test/setup_worker_test.exs:
```diff
@@ -105,6 +106,7 @@ defmodule RepoPoller.SetupWorkerTest do
              wait_for(fn ->
                %{workers: num} = DynamicSupervisor.count_children(PollerSupervisor)
                num == 1
+               :timer.sleep 200
```
i think, instead of doing a timer sleep 200 here, is better to increase the waiting time `def wait_for(timeout \\ 1000, f)` so we can increase it to 2sec or more if necessary

sescobb27 commented on apps/bugs_bunny/lib/bugs_bunny/worker/rabbit_connection.ex:
```diff
@@ -69,24 +70,29 @@ defmodule BugsBunny.Worker.RabbitConnection do
-    {:ok, %State{config: config}}
+
+    # split our opts from the ones passed to the amqp client
+    {opts, amqp_config} = Keyword.split(config, [:adapter])
```
TIL about split, nice

sescobb27 commented on apps/bugs_bunny/lib/bugs_bunny/worker/rabbit_connection.ex:
```diff
@@ -135,9 +137,36 @@ defmodule BugsBunny.Worker.RabbitConnection do
-    get_client(config).open_connection(config)
-    |> handle_rabbit_connect(state)
+  def handle_info(:connect, %{adapter: adapter, config: config} = state) do
+    # TODO: add exponential backoff for reconnects
```
i prefer moving this to its own method but i think is a matter of taste 👍

gausby replied:
I just moved the comment when I inlined the code :)

gausby replied:
ah, I just re-read your comment; I think you are referring to the code below your comment.

The reason I like it this way is that I prefer the return expression to be in the same block as the `handle_*` function definition. So often I call out to a `case` and pattern match on the result so I can return the `{:noreply, …}`. It is true you can have that returned from a helper function, but that will make it harder for the reader to know what is going on as the reader will have to follow the code path to figure out what will happen.

Does that make sense? Perhaps I should write a blogpost about this :)

sescobb27 reviewed (approved):
everything looks good to me, the only thing i'm not sure about are timeouts, because we did not need them before, so not sure why we need them now, can you create an issue so we can investigate that in the future? 👍

Commit:
Inline a function only referenced once

Commit:
Move logging into the functions related to the action logged

Commit:
Don't use async:true when tests use capture_log

Commit:
Coerce input and call the regular function instead of duplication

Commit:
Disable async testing and remove timeouts

sescobb27 commented on apps/bugs_bunny/lib/bugs_bunny/clients/rabbitmq.ex:
```diff
@@ -48,7 +58,11 @@ defmodule BugsBunny.RabbitMQ do
+    case Exchange.declare(channel, exchange, type, options) do
+      :ok ->
+        Logger.info("exchange #{exchange} successfully declared")
+        :ok
```
we need an `error` clause here, so, if we get an error we don't get a pattern match error instead
```ex
{:error, reason} ->
  {:error, reason}
```

sescobb27 commented on apps/bugs_bunny/lib/bugs_bunny/clients/rabbitmq.ex:
```diff
@@ -57,6 +71,10 @@ defmodule BugsBunny.RabbitMQ do
+    case Queue.bind(channel, queue, exchange, options) do
+      :ok ->
+        Logger.info("#{queue} successfully bound to #{exchange}")
+        :ok
```
same here

Commit:
Run `mix format`

Commit:
Pass on {:error, reason} from rabbitmq functions

Commit:
Update the version of Elixir to 1.7.4

Stuff is failing on Travis but not my machine, and this seems to be
the difference between the two setups.
